File: example/subscriber-example.py
"""
Subscriber Example - Listening to Data Feeds via WebSocket (Python)

This example demonstrates how to:
1. Create a subscriber API key for a subscription
2. Connect to WebSocket and listen to data feed updates
3. Handle real-time data updates

Prerequisites:
- Python 3.7+
- Install dependencies: pip install websocket-client requests
- A subscription ID (obtained after subscribing to a feed via frontend/wallet)
- Consumer address (your wallet address)
- Feed ID you want to listen to
"""

import os
import json
import time
import signal
import sys
import requests
import websocket
from typing import Optional

# Configuration
API_BASE_URL = os.getenv('API_URL', 'http://localhost:3001')
WS_URL = os.getenv('WS_URL', 'ws://localhost:3001')
SUBSCRIBER_API_KEY = os.getenv('SUBSCRIBER_API_KEY', 'sk_your_subscriber_api_key_here')
FEED_ID = os.getenv('FEED_ID', '0x_your_feed_id_here')
SUBSCRIPTION_ID = os.getenv('SUBSCRIPTION_ID', '0x_your_subscription_id_here')
CONSUMER_ADDRESS = os.getenv('CONSUMER_ADDRESS', '0x_your_consumer_address_here')

# Global WebSocket connection
ws_connection: Optional[websocket.WebSocketApp] = None


def create_subscriber_api_key(subscription_id: str, consumer_address: str) -> str:
    """Create a subscriber API key for a subscription."""
    try:
        print(f'Creating subscriber API key for subscription {subscription_id}...')
        
        response = requests.post(
            f'{API_BASE_URL}/api/api-keys/subscriber',
            json={
                'subscriptionId': subscription_id,
                'consumerAddress': consumer_address,
                'name': 'My Subscriber Key',
                'description': 'API key for accessing subscribed data feed',
            },
            headers={
                'Content-Type': 'application/json',
            }
        )
        
        response.raise_for_status()
        data = response.json()
        
        if data.get('success'):
            api_key = data['data']['apiKey']
            print('✅ Subscriber API key created!')
            print(f'API Key: {api_key}')
            print('⚠️  Save this key securely - it will not be shown again!')
            return api_key
        else:
            raise Exception(data.get('error', 'Failed to create API key'))
    except Exception as e:
        print(f'❌ Error creating API key: {e}')
        raise


def get_feed_data(feed_id: str, api_key: str) -> dict:
    """Get feed data via REST API (one-time fetch)."""
    try:
        print(f'Fetching data for feed {feed_id}...')
        
        response = requests.get(
            f'{API_BASE_URL}/api/data/{feed_id}',
            headers={
                'X-API-Key': api_key,
            }
        )
        
        response.raise_for_status()
        data = response.json()
        
        if data.get('success'):
            print('✅ Data retrieved successfully!')
            print('Data:', json.dumps(data['data'], indent=2))
            return data['data']
        else:
            raise Exception(data.get('error', 'Failed to retrieve data'))
    except Exception as e:
        print(f'❌ Error fetching data: {e}')
        raise


def on_message(ws, message):
    """Handle WebSocket messages."""
    try:
        data = json.loads(message)
        msg_type = data.get('type')
        
        if msg_type == 'subscribed':
            print(f"✅ Successfully subscribed to feed: {data.get('feedId')}")
        elif msg_type == 'data':
            print('\n📊 Received data update:')
            print(f"Feed ID: {data.get('feedId')}")
            print(f"Timestamp: {time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(data.get('timestamp') / 1000))}")
            print('Data:', json.dumps(data.get('data'), indent=2))
        elif msg_type == 'error':
            print(f"❌ WebSocket error: {data.get('error')}")
        elif msg_type == 'unsubscribed':
            print('✅ Unsubscribed from feed')
        else:
            print(f'📨 Received message: {data}')
    except Exception as e:
        print(f'❌ Error parsing message: {e}')


def on_error(ws, error):
    """Handle WebSocket errors."""
    print(f'❌ WebSocket error: {error}')


def on_close(ws, close_status_code, close_msg):
    """Handle WebSocket close."""
    print(f'🔌 WebSocket closed: {close_status_code} - {close_msg}')


def on_open(ws):
    """Handle WebSocket open."""
    print('✅ WebSocket connected')
    
    # Subscribe to the feed using API key
    subscribe_message = {
        'type': 'subscribe',
        'feedId': FEED_ID,
        'apiKey': SUBSCRIBER_API_KEY,
    }
    
    print('📤 Sending subscribe message:', json.dumps(subscribe_message, indent=2))
    ws.send(json.dumps(subscribe_message))


def listen_to_feed(feed_id: str, api_key: str):
    """Connect to WebSocket and listen to data feed updates."""
    global ws_connection
    
    print(f'\n🔌 Connecting to WebSocket: {WS_URL}/ws')
    
    ws_connection = websocket.WebSocketApp(
        f'{WS_URL}/ws',
        on_message=on_message,
        on_error=on_error,
        on_close=on_close,
        on_open=on_open
    )
    
    # Run forever (until interrupted)
    ws_connection.run_forever()


def signal_handler(sig, frame):
    """Handle Ctrl+C gracefully."""
    global ws_connection
    print('\n🛑 Unsubscribing and closing connection...')
    
    if ws_connection:
        try:
            unsubscribe_msg = json.dumps({'type': 'unsubscribe'})
            ws_connection.send(unsubscribe_msg)
            time.sleep(1)
            ws_connection.close()
        except:
            pass
    
    sys.exit(0)


def main():
    """Main example function."""
    print('🚀 Subscriber Example - Listening to Data Feeds (Python)\n')
    print(f'API Base URL: {API_BASE_URL}')
    print(f'WebSocket URL: {WS_URL}/ws')
    print(f'Feed ID: {FEED_ID}')
    print(f'Subscription ID: {SUBSCRIPTION_ID}')
    print(f'Consumer Address: {CONSUMER_ADDRESS}')
    print('')
    
    # Register signal handler for graceful shutdown
    signal.signal(signal.SIGINT, signal_handler)
    
    try:
        # Step 1: Create a subscriber API key (if you don't have one)
        # Uncomment if you need to create a new API key:
        # api_key = create_subscriber_api_key(SUBSCRIPTION_ID, CONSUMER_ADDRESS)
        # print('')
        
        # Use existing API key or the one you just created
        api_key = SUBSCRIBER_API_KEY
        
        if not api_key or api_key == 'sk_your_subscriber_api_key_here':
            print('❌ Please set SUBSCRIBER_API_KEY environment variable or create one using create_subscriber_api_key()')
            print('   You can create an API key by uncommenting the create_subscriber_api_key() call above')
            sys.exit(1)
        
        # Step 2: Optionally fetch data once via REST API
        print('--- Fetching data via REST API ---')
        get_feed_data(FEED_ID, api_key)
        print('')
        
        # Step 3: Connect to WebSocket and listen for real-time updates
        print('--- Connecting to WebSocket for real-time updates ---')
        print('Press Ctrl+C to stop listening...\n')
        
        listen_to_feed(FEED_ID, api_key)
        
    except Exception as e:
        print(f'Example failed: {e}')
        sys.exit(1)


if __name__ == '__main__':
    main()

File: example/subscriber-example.js
/**
 * Subscriber Example - Listening to Data Feeds via WebSocket
 * 
 * This example demonstrates how to:
 * 1. Create a subscriber API key for a subscription
 * 2. Connect to WebSocket and listen to data feed updates
 * 3. Handle real-time data updates
 * 
 * Prerequisites:
 * - Node.js installed
 * - A subscription ID (obtained after subscribing to a feed via frontend/wallet)
 * - Consumer address (your wallet address)
 * - Feed ID you want to listen to
 */

const WebSocket = require('ws');
const axios = require('axios');

// Configuration
const API_BASE_URL = process.env.API_URL || 'http://localhost:3001';
const WS_URL = process.env.WS_URL || 'ws://localhost:3001';
const SUBSCRIBER_API_KEY = process.env.SUBSCRIBER_API_KEY || 'sk_your_subscriber_api_key_here';
const FEED_ID = process.env.FEED_ID || '0x_your_feed_id_here';
const SUBSCRIPTION_ID = process.env.SUBSCRIPTION_ID || '0x_your_subscription_id_here';
const CONSUMER_ADDRESS = process.env.CONSUMER_ADDRESS || '0x_your_consumer_address_here';

/**
 * Create a subscriber API key for a subscription
 * Note: You need to have already subscribed to the feed via the frontend/wallet first
 */
async function createSubscriberApiKey(subscriptionId, consumerAddress) {
  try {
    console.log(`Creating subscriber API key for subscription ${subscriptionId}...`);
    
    const response = await axios.post(
      `${API_BASE_URL}/api/api-keys/subscriber`,
      {
        subscriptionId,
        consumerAddress,
        name: 'My Subscriber Key',
        description: 'API key for accessing subscribed data feed',
      },
      {
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );

    if (response.data.success) {
      console.log('✅ Subscriber API key created!');
      console.log('API Key:', response.data.data.apiKey);
      console.log('⚠️  Save this key securely - it will not be shown again!');
      return response.data.data.apiKey;
    } else {
      throw new Error(response.data.error || 'Failed to create API key');
    }
  } catch (error) {
    console.error('❌ Error creating API key:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Get feed data via REST API (one-time fetch)
 */
async function getFeedData(feedId, apiKey) {
  try {
    console.log(`Fetching data for feed ${feedId}...`);
    
    const response = await axios.get(`${API_BASE_URL}/api/data/${feedId}`, {
      headers: {
        'X-API-Key': apiKey,
      },
    });

    if (response.data.success) {
      console.log('✅ Data retrieved successfully!');
      console.log('Data:', JSON.stringify(response.data.data, null, 2));
      return response.data.data;
    } else {
      throw new Error(response.data.error || 'Failed to retrieve data');
    }
  } catch (error) {
    console.error('❌ Error fetching data:', error.response?.data || error.message);
    throw error;
  }
}

/**
 * Connect to WebSocket and listen to data feed updates
 */
function listenToFeed(feedId, apiKey) {
  return new Promise((resolve, reject) => {
    console.log(`\n🔌 Connecting to WebSocket: ${WS_URL}/ws`);
    
    const ws = new WebSocket(`${WS_URL}/ws`);

    ws.on('open', () => {
      console.log('✅ WebSocket connected');
      
      // Subscribe to the feed using API key
      const subscribeMessage = {
        type: 'subscribe',
        feedId: feedId,
        apiKey: apiKey,
      };
      
      console.log('📤 Sending subscribe message:', JSON.stringify(subscribeMessage, null, 2));
      ws.send(JSON.stringify(subscribeMessage));
    });

    ws.on('message', (data) => {
      try {
        const message = JSON.parse(data.toString());
        
        switch (message.type) {
          case 'subscribed':
            console.log('✅ Successfully subscribed to feed:', message.feedId);
            break;
            
          case 'data':
            console.log('\n📊 Received data update:');
            console.log('Feed ID:', message.feedId);
            console.log('Timestamp:', new Date(message.timestamp).toISOString());
            console.log('Data:', JSON.stringify(message.data, null, 2));
            break;
            
          case 'error':
            console.error('❌ WebSocket error:', message.error);
            reject(new Error(message.error));
            break;
            
          case 'unsubscribed':
            console.log('✅ Unsubscribed from feed');
            break;
            
          default:
            console.log('📨 Received message:', message);
        }
      } catch (error) {
        console.error('❌ Error parsing message:', error);
      }
    });

    ws.on('error', (error) => {
      console.error('❌ WebSocket error:', error);
      reject(error);
    });

    ws.on('close', (code, reason) => {
      console.log(`🔌 WebSocket closed: ${code} - ${reason}`);
      resolve();
    });

    // Handle graceful shutdown
    process.on('SIGINT', () => {
      console.log('\n🛑 Unsubscribing and closing connection...');
      ws.send(JSON.stringify({ type: 'unsubscribe' }));
      setTimeout(() => {
        ws.close();
        process.exit(0);
      }, 1000);
    });
  });
}

/**
 * Main example function
 */
async function main() {
  console.log('🚀 Subscriber Example - Listening to Data Feeds\n');
  console.log('API Base URL:', API_BASE_URL);
  console.log('WebSocket URL:', `${WS_URL}/ws`);
  console.log('Feed ID:', FEED_ID);
  console.log('Subscription ID:', SUBSCRIPTION_ID);
  console.log('Consumer Address:', CONSUMER_ADDRESS);
  console.log('');

  try {
    // Step 1: Create a subscriber API key (if you don't have one)
    // Uncomment if you need to create a new API key:
    // const apiKey = await createSubscriberApiKey(SUBSCRIPTION_ID, CONSUMER_ADDRESS);
    // console.log('');
    
    // Use existing API key or the one you just created
    const apiKey = SUBSCRIBER_API_KEY;
    
    if (!apiKey || apiKey === 'sk_your_subscriber_api_key_here') {
      console.error('❌ Please set SUBSCRIBER_API_KEY environment variable or create one using createSubscriberApiKey()');
      console.error('   You can create an API key by uncommenting the createSubscriberApiKey() call above');
      process.exit(1);
    }

    // Step 2: Optionally fetch data once via REST API
    console.log('--- Fetching data via REST API ---');
    await getFeedData(FEED_ID, apiKey);
    console.log('');

    // Step 3: Connect to WebSocket and listen for real-time updates
    console.log('--- Connecting to WebSocket for real-time updates ---');
    console.log('Press Ctrl+C to stop listening...\n');
    
    await listenToFeed(FEED_ID, apiKey);

  } catch (error) {
    console.error('Example failed:', error.message);
    process.exit(1);
  }
}

// Run example if executed directly
if (require.main === module) {
  main().catch(console.error);
}

module.exports = {
  createSubscriberApiKey,
  getFeedData,
  listenToFeed,
};

